File: study/examples/Algorithm/c/22_string_algorithm.c
/*
 * 문자열 알고리즘 (String Algorithms)
 * KMP, Rabin-Karp, Z-알고리즘, 매나커
 *
 * 패턴 매칭과 문자열 처리 알고리즘입니다.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_LEN 100001
#define MOD 1000000007
#define BASE 31

/* =============================================================================
 * 1. KMP (Knuth-Morris-Pratt)
 * ============================================================================= */

/* 실패 함수 (부분 일치 테이블) 계산 */
int* compute_failure(const char* pattern) {
    int m = strlen(pattern);
    int* fail = calloc(m, sizeof(int));

    int j = 0;
    for (int i = 1; i < m; i++) {
        while (j > 0 && pattern[i] != pattern[j]) {
            j = fail[j - 1];
        }
        if (pattern[i] == pattern[j]) {
            fail[i] = ++j;
        }
    }

    return fail;
}

/* KMP 검색 - 모든 매칭 위치 반환 */
int* kmp_search(const char* text, const char* pattern, int* match_count) {
    int n = strlen(text);
    int m = strlen(pattern);
    int* fail = compute_failure(pattern);
    int* matches = malloc(n * sizeof(int));
    *match_count = 0;

    int j = 0;
    for (int i = 0; i < n; i++) {
        while (j > 0 && text[i] != pattern[j]) {
            j = fail[j - 1];
        }
        if (text[i] == pattern[j]) {
            j++;
            if (j == m) {
                matches[(*match_count)++] = i - m + 1;
                j = fail[j - 1];
            }
        }
    }

    free(fail);
    return matches;
}

/* =============================================================================
 * 2. Rabin-Karp
 * ============================================================================= */

long long compute_hash(const char* s, int len) {
    long long hash = 0;
    long long power = 1;
    for (int i = 0; i < len; i++) {
        hash = (hash + (s[i] - 'a' + 1) * power) % MOD;
        power = (power * BASE) % MOD;
    }
    return hash;
}

int* rabin_karp(const char* text, const char* pattern, int* match_count) {
    int n = strlen(text);
    int m = strlen(pattern);
    int* matches = malloc(n * sizeof(int));
    *match_count = 0;

    if (m > n) return matches;

    long long pattern_hash = compute_hash(pattern, m);
    long long text_hash = compute_hash(text, m);
    long long power = 1;

    for (int i = 0; i < m - 1; i++) {
        power = (power * BASE) % MOD;
    }

    for (int i = 0; i <= n - m; i++) {
        if (text_hash == pattern_hash) {
            /* 해시 충돌 확인 */
            int match = 1;
            for (int j = 0; j < m; j++) {
                if (text[i + j] != pattern[j]) {
                    match = 0;
                    break;
                }
            }
            if (match) matches[(*match_count)++] = i;
        }

        if (i < n - m) {
            /* 롤링 해시 */
            text_hash = (text_hash - (text[i] - 'a' + 1) + MOD) % MOD;
            text_hash = (text_hash * mod_inv(BASE, MOD)) % MOD;
            text_hash = (text_hash + (text[i + m] - 'a' + 1) * power) % MOD;
        }
    }

    return matches;
}

/* 모듈러 역원 (간단 버전) */
long long mod_inv(long long a, long long mod) {
    long long result = 1;
    long long exp = mod - 2;
    while (exp > 0) {
        if (exp & 1) result = (result * a) % mod;
        a = (a * a) % mod;
        exp >>= 1;
    }
    return result;
}

/* 간단한 Rabin-Karp (롤링 해시 개선) */
int* rabin_karp_simple(const char* text, const char* pattern, int* match_count) {
    int n = strlen(text);
    int m = strlen(pattern);
    int* matches = malloc(n * sizeof(int));
    *match_count = 0;

    if (m > n) return matches;

    long long pattern_hash = 0;
    long long text_hash = 0;
    long long power = 1;

    /* 초기 해시 계산 */
    for (int i = 0; i < m; i++) {
        pattern_hash = (pattern_hash * BASE + pattern[i]) % MOD;
        text_hash = (text_hash * BASE + text[i]) % MOD;
        if (i < m - 1) power = (power * BASE) % MOD;
    }

    for (int i = 0; i <= n - m; i++) {
        if (text_hash == pattern_hash) {
            if (strncmp(text + i, pattern, m) == 0) {
                matches[(*match_count)++] = i;
            }
        }
        if (i < n - m) {
            text_hash = ((text_hash - text[i] * power % MOD + MOD) * BASE + text[i + m]) % MOD;
        }
    }

    return matches;
}

/* =============================================================================
 * 3. Z-알고리즘
 * ============================================================================= */

int* z_function(const char* s) {
    int n = strlen(s);
    int* z = calloc(n, sizeof(int));
    int l = 0, r = 0;

    for (int i = 1; i < n; i++) {
        if (i < r) {
            z[i] = (z[i - l] < r - i) ? z[i - l] : r - i;
        }
        while (i + z[i] < n && s[z[i]] == s[i + z[i]]) {
            z[i]++;
        }
        if (i + z[i] > r) {
            l = i;
            r = i + z[i];
        }
    }

    return z;
}

/* Z-알고리즘으로 패턴 매칭 */
int* z_search(const char* text, const char* pattern, int* match_count) {
    int n = strlen(text);
    int m = strlen(pattern);

    /* pattern + "$" + text 연결 */
    char* combined = malloc(n + m + 2);
    sprintf(combined, "%s$%s", pattern, text);

    int* z = z_function(combined);
    int* matches = malloc(n * sizeof(int));
    *match_count = 0;

    for (int i = m + 1; i < n + m + 1; i++) {
        if (z[i] == m) {
            matches[(*match_count)++] = i - m - 1;
        }
    }

    free(combined);
    free(z);
    return matches;
}

/* =============================================================================
 * 4. 매나커 알고리즘 (Manacher)
 * ============================================================================= */

/* 가장 긴 팰린드롬 부분문자열 */
char* manacher(const char* s) {
    int n = strlen(s);
    if (n == 0) return strdup("");

    /* 문자 사이에 # 삽입 */
    int len = 2 * n + 1;
    char* t = malloc(len + 1);
    for (int i = 0; i < n; i++) {
        t[2 * i] = '#';
        t[2 * i + 1] = s[i];
    }
    t[len - 1] = '#';
    t[len] = '\0';

    int* p = calloc(len, sizeof(int));
    int center = 0, right = 0;
    int max_len = 0, max_center = 0;

    for (int i = 0; i < len; i++) {
        if (i < right) {
            int mirror = 2 * center - i;
            p[i] = (p[mirror] < right - i) ? p[mirror] : right - i;
        }

        while (i - p[i] - 1 >= 0 && i + p[i] + 1 < len &&
               t[i - p[i] - 1] == t[i + p[i] + 1]) {
            p[i]++;
        }

        if (i + p[i] > right) {
            center = i;
            right = i + p[i];
        }

        if (p[i] > max_len) {
            max_len = p[i];
            max_center = i;
        }
    }

    /* 원본 문자열에서 추출 */
    int start = (max_center - max_len) / 2;
    char* result = malloc(max_len + 1);
    strncpy(result, s + start, max_len);
    result[max_len] = '\0';

    free(t);
    free(p);
    return result;
}

/* =============================================================================
 * 5. 문자열 해싱
 * ============================================================================= */

typedef struct {
    long long* prefix_hash;
    long long* power;
    int len;
} StringHash;

StringHash* create_string_hash(const char* s) {
    int n = strlen(s);
    StringHash* sh = malloc(sizeof(StringHash));
    sh->prefix_hash = malloc((n + 1) * sizeof(long long));
    sh->power = malloc((n + 1) * sizeof(long long));
    sh->len = n;

    sh->prefix_hash[0] = 0;
    sh->power[0] = 1;

    for (int i = 0; i < n; i++) {
        sh->prefix_hash[i + 1] = (sh->prefix_hash[i] * BASE + s[i]) % MOD;
        sh->power[i + 1] = (sh->power[i] * BASE) % MOD;
    }

    return sh;
}

long long get_hash(StringHash* sh, int l, int r) {
    long long hash = (sh->prefix_hash[r + 1] -
                      sh->prefix_hash[l] * sh->power[r - l + 1] % MOD + MOD) % MOD;
    return hash;
}

void free_string_hash(StringHash* sh) {
    free(sh->prefix_hash);
    free(sh->power);
    free(sh);
}

/* =============================================================================
 * 6. 유용한 문자열 함수들
 * ============================================================================= */

/* 접미사 배열 (간단 버전, O(n² log n)) */
int* suffix_array_simple(const char* s) {
    int n = strlen(s);
    int* sa = malloc(n * sizeof(int));
    for (int i = 0; i < n; i++) sa[i] = i;

    /* 정렬 */
    for (int i = 0; i < n - 1; i++) {
        for (int j = i + 1; j < n; j++) {
            if (strcmp(s + sa[i], s + sa[j]) > 0) {
                int temp = sa[i];
                sa[i] = sa[j];
                sa[j] = temp;
            }
        }
    }

    return sa;
}

/* LCP 배열 (Kasai 알고리즘) */
int* lcp_array(const char* s, int* sa) {
    int n = strlen(s);
    int* rank = malloc(n * sizeof(int));
    int* lcp = malloc(n * sizeof(int));

    for (int i = 0; i < n; i++) rank[sa[i]] = i;

    int k = 0;
    for (int i = 0; i < n; i++) {
        if (rank[i] == 0) {
            lcp[0] = 0;
            continue;
        }
        int j = sa[rank[i] - 1];
        while (i + k < n && j + k < n && s[i + k] == s[j + k]) k++;
        lcp[rank[i]] = k;
        if (k > 0) k--;
    }

    free(rank);
    return lcp;
}

/* =============================================================================
 * 테스트
 * ============================================================================= */

int main(void) {
    printf("============================================================\n");
    printf("문자열 알고리즘 예제\n");
    printf("============================================================\n");

    /* 1. KMP */
    printf("\n[1] KMP 알고리즘\n");
    const char* text1 = "ABABDABACDABABCABAB";
    const char* pattern1 = "ABABCABAB";
    int match_count;
    int* matches = kmp_search(text1, pattern1, &match_count);
    printf("    텍스트: %s\n", text1);
    printf("    패턴: %s\n", pattern1);
    printf("    매칭 위치 (%d개): ", match_count);
    for (int i = 0; i < match_count; i++) printf("%d ", matches[i]);
    printf("\n");
    free(matches);

    /* 실패 함수 시각화 */
    const char* pattern2 = "ABAABAB";
    int* fail = compute_failure(pattern2);
    printf("    패턴 \"%s\"의 실패 함수: ", pattern2);
    for (int i = 0; i < (int)strlen(pattern2); i++) printf("%d ", fail[i]);
    printf("\n");
    free(fail);

    /* 2. Z-알고리즘 */
    printf("\n[2] Z-알고리즘\n");
    const char* str = "aabxaabxcaabxaabxay";
    int* z = z_function(str);
    printf("    문자열: %s\n", str);
    printf("    Z-배열: ");
    for (int i = 0; i < (int)strlen(str); i++) printf("%d ", z[i]);
    printf("\n");
    free(z);

    matches = z_search(text1, pattern1, &match_count);
    printf("    Z-알고리즘 매칭 위치: ");
    for (int i = 0; i < match_count; i++) printf("%d ", matches[i]);
    printf("\n");
    free(matches);

    /* 3. Rabin-Karp */
    printf("\n[3] Rabin-Karp\n");
    const char* text2 = "abracadabra";
    const char* pattern3 = "abra";
    matches = rabin_karp_simple(text2, pattern3, &match_count);
    printf("    텍스트: %s\n", text2);
    printf("    패턴: %s\n", pattern3);
    printf("    매칭 위치: ");
    for (int i = 0; i < match_count; i++) printf("%d ", matches[i]);
    printf("\n");
    free(matches);

    /* 4. 매나커 */
    printf("\n[4] 매나커 알고리즘\n");
    const char* str2 = "babad";
    char* palindrome = manacher(str2);
    printf("    문자열: %s\n", str2);
    printf("    가장 긴 팰린드롬: %s\n", palindrome);
    free(palindrome);

    const char* str3 = "forgeeksskeegfor";
    palindrome = manacher(str3);
    printf("    문자열: %s\n", str3);
    printf("    가장 긴 팰린드롬: %s\n", palindrome);
    free(palindrome);

    /* 5. 문자열 해싱 */
    printf("\n[5] 문자열 해싱\n");
    const char* str4 = "abcabc";
    StringHash* sh = create_string_hash(str4);
    printf("    문자열: %s\n", str4);
    printf("    hash[0:2] = %lld\n", get_hash(sh, 0, 2));
    printf("    hash[3:5] = %lld\n", get_hash(sh, 3, 5));
    printf("    동일 여부: %s\n",
           get_hash(sh, 0, 2) == get_hash(sh, 3, 5) ? "예" : "아니오");
    free_string_hash(sh);

    /* 6. 접미사 배열 */
    printf("\n[6] 접미사 배열\n");
    const char* str5 = "banana";
    int* sa = suffix_array_simple(str5);
    printf("    문자열: %s\n", str5);
    printf("    접미사 배열: ");
    for (int i = 0; i < (int)strlen(str5); i++) printf("%d ", sa[i]);
    printf("\n");
    printf("    정렬된 접미사:\n");
    for (int i = 0; i < (int)strlen(str5); i++) {
        printf("      %d: %s\n", sa[i], str5 + sa[i]);
    }
    free(sa);

    /* 7. 복잡도 */
    printf("\n[7] 복잡도\n");
    printf("    | 알고리즘        | 전처리    | 검색      |\n");
    printf("    |-----------------|-----------|----------|\n");
    printf("    | KMP             | O(m)      | O(n)     |\n");
    printf("    | Rabin-Karp      | O(m)      | O(n+m)*  |\n");
    printf("    | Z-알고리즘      | O(n+m)    | -        |\n");
    printf("    | 매나커          | -         | O(n)     |\n");
    printf("    * 평균, 최악 O(nm)\n");

    printf("\n============================================================\n");

    return 0;
}

File: study/examples/Algorithm/python/22_string_algorithm.py
"""
문자열 알고리즘 (String Algorithms)
String Matching and Processing

문자열 검색 및 처리 알고리즘을 구현합니다.
"""

from typing import List, Tuple


# =============================================================================
# 1. KMP 알고리즘 (Knuth-Morris-Pratt)
# =============================================================================

def kmp_failure(pattern: str) -> List[int]:
    """
    KMP 실패 함수 (부분 일치 테이블) 계산
    시간복잡도: O(m), m = 패턴 길이
    """
    m = len(pattern)
    failure = [0] * m
    j = 0  # 이전 최대 접두사 길이

    for i in range(1, m):
        while j > 0 and pattern[i] != pattern[j]:
            j = failure[j - 1]

        if pattern[i] == pattern[j]:
            j += 1
            failure[i] = j

    return failure


def kmp_search(text: str, pattern: str) -> List[int]:
    """
    KMP 문자열 검색
    시간복잡도: O(n + m)
    반환: 패턴이 발견된 시작 인덱스 리스트
    """
    if not pattern:
        return []

    n, m = len(text), len(pattern)
    failure = kmp_failure(pattern)
    matches = []
    j = 0  # 패턴 인덱스

    for i in range(n):
        while j > 0 and text[i] != pattern[j]:
            j = failure[j - 1]

        if text[i] == pattern[j]:
            if j == m - 1:
                matches.append(i - m + 1)
                j = failure[j]
            else:
                j += 1

    return matches


# =============================================================================
# 2. Rabin-Karp 알고리즘
# =============================================================================

def rabin_karp_search(text: str, pattern: str, mod: int = 10**9 + 7) -> List[int]:
    """
    Rabin-Karp 문자열 검색 (롤링 해시)
    시간복잡도: 평균 O(n + m), 최악 O(nm)
    """
    if not pattern or len(pattern) > len(text):
        return []

    n, m = len(text), len(pattern)
    base = 256
    matches = []

    # 패턴 해시 계산
    pattern_hash = 0
    text_hash = 0
    h = pow(base, m - 1, mod)

    for i in range(m):
        pattern_hash = (pattern_hash * base + ord(pattern[i])) % mod
        text_hash = (text_hash * base + ord(text[i])) % mod

    for i in range(n - m + 1):
        if pattern_hash == text_hash:
            # 해시 충돌 확인
            if text[i:i + m] == pattern:
                matches.append(i)

        # 다음 윈도우 해시 계산
        if i < n - m:
            text_hash = ((text_hash - ord(text[i]) * h) * base + ord(text[i + m])) % mod

    return matches


# =============================================================================
# 3. Z 알고리즘
# =============================================================================

def z_function(s: str) -> List[int]:
    """
    Z 함수 계산
    z[i] = s와 s[i:]의 최장 공통 접두사 길이
    시간복잡도: O(n)
    """
    n = len(s)
    z = [0] * n
    z[0] = n

    l, r = 0, 0  # Z-box [l, r)

    for i in range(1, n):
        if i < r:
            z[i] = min(r - i, z[i - l])

        while i + z[i] < n and s[z[i]] == s[i + z[i]]:
            z[i] += 1

        if i + z[i] > r:
            l, r = i, i + z[i]

    return z


def z_search(text: str, pattern: str) -> List[int]:
    """Z 알고리즘을 이용한 문자열 검색"""
    if not pattern:
        return []

    combined = pattern + "$" + text
    z = z_function(combined)
    m = len(pattern)

    return [i - m - 1 for i in range(m + 1, len(combined)) if z[i] == m]


# =============================================================================
# 4. Manacher 알고리즘 (최장 회문 부분문자열)
# =============================================================================

def manacher(s: str) -> Tuple[int, int]:
    """
    최장 회문 부분문자열 찾기
    시간복잡도: O(n)
    반환: (시작 인덱스, 길이)
    """
    if not s:
        return 0, 0

    # 전처리: 문자 사이에 # 삽입
    t = '#' + '#'.join(s) + '#'
    n = len(t)
    p = [0] * n  # p[i] = i 중심 회문 반지름

    c, r = 0, 0  # 현재 회문 중심, 오른쪽 경계

    for i in range(n):
        if i < r:
            p[i] = min(r - i, p[2 * c - i])

        # 확장 시도
        while i - p[i] - 1 >= 0 and i + p[i] + 1 < n and t[i - p[i] - 1] == t[i + p[i] + 1]:
            p[i] += 1

        # 경계 업데이트
        if i + p[i] > r:
            c, r = i, i + p[i]

    # 최장 회문 찾기
    max_len = max(p)
    center = p.index(max_len)

    # 원본 문자열에서의 위치
    start = (center - max_len) // 2
    length = max_len

    return start, length


def longest_palindrome(s: str) -> str:
    """최장 회문 부분문자열 반환"""
    start, length = manacher(s)
    return s[start:start + length]


# =============================================================================
# 5. 접미사 배열 (Suffix Array) - 간단 구현
# =============================================================================

def suffix_array(s: str) -> List[int]:
    """
    접미사 배열 생성 (간단한 O(n log² n) 구현)
    반환: 사전순 정렬된 접미사의 시작 인덱스
    """
    n = len(s)
    sa = list(range(n))
    rank = [ord(c) for c in s]
    tmp = [0] * n

    k = 1
    while k < n:
        # (rank[i], rank[i+k]) 기준 정렬
        def key(i):
            return (rank[i], rank[i + k] if i + k < n else -1)

        sa.sort(key=key)

        # 새 rank 계산
        tmp[sa[0]] = 0
        for i in range(1, n):
            tmp[sa[i]] = tmp[sa[i - 1]]
            if key(sa[i]) != key(sa[i - 1]):
                tmp[sa[i]] += 1

        rank = tmp[:]
        k *= 2

    return sa


def lcp_array(s: str, sa: List[int]) -> List[int]:
    """
    LCP 배열 (Longest Common Prefix)
    lcp[i] = s[sa[i]:]와 s[sa[i+1]:]의 최장 공통 접두사 길이
    시간복잡도: O(n)
    """
    n = len(s)
    rank = [0] * n
    for i, idx in enumerate(sa):
        rank[idx] = i

    lcp = [0] * (n - 1)
    h = 0

    for i in range(n):
        if rank[i] > 0:
            j = sa[rank[i] - 1]
            while i + h < n and j + h < n and s[i + h] == s[j + h]:
                h += 1
            lcp[rank[i] - 1] = h
            if h > 0:
                h -= 1

    return lcp


# =============================================================================
# 6. 트라이 기반 문자열 검색
# =============================================================================

class TrieNode:
    def __init__(self):
        self.children = {}
        self.is_end = False
        self.output = []  # Aho-Corasick용


class AhoCorasick:
    """
    Aho-Corasick 알고리즘 (다중 패턴 검색)
    전처리: O(Σ|patterns|)
    검색: O(n + m), m = 매칭 수
    """

    def __init__(self, patterns: List[str]):
        self.root = TrieNode()
        self.patterns = patterns
        self._build_trie()
        self._build_failure()

    def _build_trie(self):
        for idx, pattern in enumerate(self.patterns):
            node = self.root
            for char in pattern:
                if char not in node.children:
                    node.children[char] = TrieNode()
                node = node.children[char]
            node.is_end = True
            node.output.append(idx)

    def _build_failure(self):
        from collections import deque

        queue = deque()
        self.root.fail = self.root

        for child in self.root.children.values():
            child.fail = self.root
            queue.append(child)

        while queue:
            node = queue.popleft()

            for char, child in node.children.items():
                fail = node.fail
                while fail != self.root and char not in fail.children:
                    fail = fail.fail

                child.fail = fail.children.get(char, self.root)
                if child.fail == child:
                    child.fail = self.root

                child.output += child.fail.output
                queue.append(child)

    def search(self, text: str) -> List[Tuple[int, int]]:
        """
        텍스트에서 모든 패턴 검색
        반환: [(위치, 패턴 인덱스), ...]
        """
        results = []
        node = self.root

        for i, char in enumerate(text):
            while node != self.root and char not in node.children:
                node = node.fail

            node = node.children.get(char, self.root)

            for pattern_idx in node.output:
                pattern = self.patterns[pattern_idx]
                results.append((i - len(pattern) + 1, pattern_idx))

        return results


# =============================================================================
# 7. 편집 거리 (Edit Distance)
# =============================================================================

def edit_distance(s1: str, s2: str) -> int:
    """
    레벤슈타인 거리 (편집 거리)
    시간복잡도: O(mn)
    공간복잡도: O(min(m, n)) 최적화 가능
    """
    m, n = len(s1), len(s2)

    # 공간 최적화: 두 행만 사용
    prev = list(range(n + 1))
    curr = [0] * (n + 1)

    for i in range(1, m + 1):
        curr[0] = i

        for j in range(1, n + 1):
            if s1[i - 1] == s2[j - 1]:
                curr[j] = prev[j - 1]
            else:
                curr[j] = 1 + min(prev[j - 1], prev[j], curr[j - 1])

        prev, curr = curr, prev

    return prev[n]


# =============================================================================
# 8. 문자열 해싱
# =============================================================================

class StringHash:
    """
    다항식 롤링 해시
    충돌을 줄이기 위해 두 개의 해시 사용 (double hashing)
    """

    def __init__(self, s: str):
        self.s = s
        self.n = len(s)
        self.MOD1 = 10**9 + 7
        self.MOD2 = 10**9 + 9
        self.BASE1 = 31
        self.BASE2 = 37

        self.hash1 = [0] * (self.n + 1)
        self.hash2 = [0] * (self.n + 1)
        self.pow1 = [1] * (self.n + 1)
        self.pow2 = [1] * (self.n + 1)

        for i in range(self.n):
            self.hash1[i + 1] = (self.hash1[i] * self.BASE1 + ord(s[i])) % self.MOD1
            self.hash2[i + 1] = (self.hash2[i] * self.BASE2 + ord(s[i])) % self.MOD2
            self.pow1[i + 1] = self.pow1[i] * self.BASE1 % self.MOD1
            self.pow2[i + 1] = self.pow2[i] * self.BASE2 % self.MOD2

    def get_hash(self, l: int, r: int) -> Tuple[int, int]:
        """s[l:r]의 해시 값 (0-indexed, 반열린 구간)"""
        h1 = (self.hash1[r] - self.hash1[l] * self.pow1[r - l]) % self.MOD1
        h2 = (self.hash2[r] - self.hash2[l] * self.pow2[r - l]) % self.MOD2
        return (h1, h2)

    def is_equal(self, l1: int, r1: int, l2: int, r2: int) -> bool:
        """두 부분문자열이 같은지 확인"""
        if r1 - l1 != r2 - l2:
            return False
        return self.get_hash(l1, r1) == self.get_hash(l2, r2)


# =============================================================================
# 테스트
# =============================================================================

def main():
    print("=" * 60)
    print("문자열 알고리즘 (String Algorithms) 예제")
    print("=" * 60)

    # 1. KMP
    print("\n[1] KMP 알고리즘")
    text = "ABABDABACDABABCABAB"
    pattern = "ABABCABAB"
    failure = kmp_failure(pattern)
    matches = kmp_search(text, pattern)
    print(f"    텍스트: {text}")
    print(f"    패턴: {pattern}")
    print(f"    실패 함수: {failure}")
    print(f"    매칭 위치: {matches}")

    # 2. Rabin-Karp
    print("\n[2] Rabin-Karp 알고리즘")
    matches = rabin_karp_search(text, pattern)
    print(f"    매칭 위치: {matches}")

    # 3. Z 알고리즘
    print("\n[3] Z 알고리즘")
    s = "aabxaab"
    z = z_function(s)
    print(f"    문자열: {s}")
    print(f"    Z 배열: {z}")
    matches = z_search(text, pattern)
    print(f"    검색 결과: {matches}")

    # 4. Manacher
    print("\n[4] Manacher 알고리즘 (최장 회문)")
    s = "babad"
    palindrome = longest_palindrome(s)
    print(f"    문자열: {s}")
    print(f"    최장 회문: {palindrome}")

    s2 = "abacdfgdcaba"
    palindrome2 = longest_palindrome(s2)
    print(f"    문자열: {s2}")
    print(f"    최장 회문: {palindrome2}")

    # 5. 접미사 배열
    print("\n[5] 접미사 배열")
    s = "banana"
    sa = suffix_array(s)
    lcp = lcp_array(s, sa)
    print(f"    문자열: {s}")
    print(f"    접미사 배열: {sa}")
    print("    접미사들:")
    for i in sa:
        print(f"      {i}: {s[i:]}")
    print(f"    LCP 배열: {lcp}")

    # 6. Aho-Corasick
    print("\n[6] Aho-Corasick (다중 패턴)")
    patterns = ["he", "she", "his", "hers"]
    text = "ahishers"
    ac = AhoCorasick(patterns)
    results = ac.search(text)
    print(f"    패턴: {patterns}")
    print(f"    텍스트: {text}")
    print("    매칭:")
    for pos, idx in results:
        print(f"      위치 {pos}: '{patterns[idx]}'")

    # 7. 편집 거리
    print("\n[7] 편집 거리")
    s1, s2 = "kitten", "sitting"
    dist = edit_distance(s1, s2)
    print(f"    '{s1}' → '{s2}'")
    print(f"    편집 거리: {dist}")

    # 8. 문자열 해싱
    print("\n[8] 문자열 해싱")
    s = "abcabc"
    sh = StringHash(s)
    print(f"    문자열: {s}")
    print(f"    hash(0:3) = {sh.get_hash(0, 3)}")
    print(f"    hash(3:6) = {sh.get_hash(3, 6)}")
    print(f"    s[0:3] == s[3:6]: {sh.is_equal(0, 3, 3, 6)}")

    print("\n" + "=" * 60)


if __name__ == "__main__":
    main()
